### DISCUSSION TITLE
默认1安装出错 不会弄了

### DISCUSSION THREAD

**Discussion by eerilqin** [kkkgo/PaoPaoDNS#46] [问答和帮助] [answered]
docker: Error response from daemon: driver failed programming external connectivity on endpoint paopaodns (a6e0c905eff88b0df30d3b0e7aade82675df4f1e9e1ea392e02ab70acfcf3c92): Error starting userland proxy: listen tcp4 0.0.0.0:53: bind: address already in use.
试了N遍 小白求教

**Comment by kkkgo** [comment]
由于缺乏部署过程的信息，我只能根据你的信息来猜测可能的错误原因。listen tcp4 0.0.0.0:53: bind: address already in use.   这个错误的意思是因为53端口被占用而无法启动，说明你的宿主机上有程序在占用53端口，所以无法映射53端口到容器，一般可能是宿主机上自带了DNS服务导致占用了53端口，你可以考虑找到占用53端口的程序停止相应的服务，或者使用macvlan等网络模式来部署，避免和宿主机53端口占用冲突。

**Reply by eerilqin** [reply]
有没有联系方式 不行你给我安装一下 十分感谢，我各种方法都试过 软路由固件也换了N次 还是搞不定

**Reply by kkkgo** [reply]
虽然不推荐all in boom，毕竟在软路由的系统上跑docker并不是一个好的选择。但如果你要安装在软路由上面跑的话，软路由本身肯定有DNS服务会占用53端口，你可以用macvlan的网络模式，或者修改路由器的DNS服务器端口或者关闭DNS服务。或者你可以把该docker的DNS作为你软路由的上游来使用，映射其他端口。比如你可以映射5353端口到容器的53端口：
```shell
docker run -d \
--name paopaodns \
-v /home/mydata:/data \
-e CNAUTO=yes \
--restart always \
-p 5353:53/tcp -p 5353:53/udp \
sliamb/paopaodns
```
这样你可以在软路由上设置上游为127.0.0.01:5353。具体设置的地方根据你的系统而异。比如你同时用了adguardhome的话，在DNS设置里面就可以直接指定，或者dnsmasq等也有相关设置。
当然，最理想的拓扑是主路由是只当路由器使用，docker和虚拟机由另一台server来实现。

**Reply by eerilqin** [reply, marked as answer]
stdout: =====PaoPaoDNS docker start=====
stdout: images build time : 2023-09-19 13:43:08 UTC
stdout: New version install ! Try clean...
stdout: ulimit adbove 1000000.
stdout:               total        used        free      shared  buff/cache   available
stdout: Mem:           1980         724         877           3         379        1189
stdout: Swap:             0           0           0
stdout:               total        used        free      shared  buff/cache   available
stdout: Mem:           1.9G      724.5M      876.6M        3.5M      379.1M        1.2G
stdout: Swap:             0           0           0
stdout: MEMSIZE:1189
stdout: [Warning] use safemem!
stdout: ====ENV TEST====
stdout: MEM:50m 100m 200 100mb
stdout: prefPC:9
stdout: CORES:-1-
stdout: POWCORES:-1-
stdout: ulimit :-1048576-
stdout: FDLIM :-1-
stdout: TZ:-Asia/Shanghai-
stdout: UPDATE:-weekly-
stdout: DNS_SERVERNAME:-PaoPaoDNS,blog.03k.org-
stdout: SERVER_IP:-none-
stdout: ETHIP:-172.17.0.2-
stdout: DNSPORT:-53-
stdout: SOCKS5:-IP:PORT-
stdout: CNAUTO:-yes-
stdout: IPV6:-no-
stdout: CNFALL:-yes-
stdout: CUSTOM_FORWARD:-IP:PORT-
stdout: AUTO_FORWARD:-no-
stdout: AUTO_FORWARD_CHECK:-yes-
stdout: USE_MARK_DATA:-no-
stdout: RULES_TTL:-0-
stdout: CN_TRACKER:-yes-
stdout: USE_HOSTS:-no-
stdout: HTTP_FILE:-no-
stdout: SAFEMODE:--
stdout: QUERY_TIME:-2000ms-
stdout: ADDINFO:-no-
stdout: PLATFORM:-Linux acc4366d097f 5.10.176 #0 SMP Fri Sep 8 01:54:31 2023 x86_64 Linux-
stdout: ====ENV TEST====
stdout: Apply trackerslist...
stderr: 2023-09-20T21:09:16.180+0800	INFO	working directory changed	{"path": "/tmp"}
stderr: 2023-09-20T21:09:16.189+0800	INFO	main config loaded	{"file": "mosdns.yaml"}
stdout: PID   USER     TIME  COMMAND
stdout:     1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh
stdout:    23 root      0:00 crond
stdout:   112 root      0:00 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.toml
stdout:   115 root      0:00 mosdns start -d /tmp -c mosdns.yaml
stdout:   136 root      0:00 unbound -c /tmp/unbound_raw.conf -p
stdout:   137 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
stdout:   138 root      0:00 {data_update.sh} /bin/sh /usr/sbin/data_update.sh
stdout:   141 root      0:00 sleep 253
stdout:   142 root      0:00 unbound -c /tmp/unbound_forward.conf -p
stdout:   147 root      0:00 ps
stdout:   148 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
stdout:   150 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
stdout: 151:C 20 Sep 2023 21:09:16.240 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
stdout: 151:C 20 Sep 2023 21:09:16.240 # systemd supervision requested or auto-detected, but Redis is compiled without libsystemd support!
stderr: Setting up watches.
stderr: Watches established.

南无啊弥陀佛 终于装上了  看看运行的对吗？

**Reply by eerilqin** [reply]
{
  "Path": "\/bin\/sh",
  "ProcessLabel": "",
  "ResolvConfPath": "\/mnt\/sata1-4\/docker\/containers\/acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015\/resolv.conf",
  "NetworkSettings": {
    "LinkLocalIPv6Address": "",
    "SandboxID": "dcc7649527a1d76e273ebed39c6ed1d042e99f806ea849b9ab460eb1987e65ed",
    "HairpinMode": false,
    "Networks": {
      "bridge": {
        "NetworkID": "bde5ccefff72e463fb835a8553542b9a980af433c43aca1d898717c4aed47790",
        "IPAddress": "172.17.0.2",
        "MacAddress": "02:42:ac:11:00:02",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "Gateway": "172.17.0.1",
        "GlobalIPv6PrefixLen": 0,
        "EndpointID": "f88660ea55890a6fc2935829cee21ae0702db568ffdaf120464e6fa3d761781e",
        "GlobalIPv6Address": ""
      }
    },
    "Ports": {
      "53\/udp": [
        {
          "HostIp": "0.0.0.0",
          "HostPort": "5335"
        },
        {
          "HostIp": "::",
          "HostPort": "5335"
        }
      ],
      "53\/tcp": [
        {
          "HostIp": "0.0.0.0",
          "HostPort": "5335"
        },
        {
          "HostIp": "::",
          "HostPort": "5335"
        }
      ]
    },
    "Bridge": "",
    "MacAddress": "02:42:ac:11:00:02",
    "IPv6Gateway": "",
    "IPPrefixLen": 16,
    "IPAddress": "172.17.0.2",
    "EndpointID": "f88660ea55890a6fc2935829cee21ae0702db568ffdaf120464e6fa3d761781e",
    "SandboxKey": "\/var\/run\/docker\/netns\/dcc7649527a1",
    "Gateway": "172.17.0.1",
    "GlobalIPv6PrefixLen": 0,
    "LinkLocalIPv6PrefixLen": 0,
    "GlobalIPv6Address": ""
  },
  "ExecIDs": [
    "09035026c442b4ad49839e50db480f83c852373ad76fedc679146889eb038d8a"
  ],
  "MountLabel": "",
  "HostsPath": "\/mnt\/sata1-4\/docker\/containers\/acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015\/hosts",
  "LogPath": "\/mnt\/sata1-4\/docker\/containers\/acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015\/acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015-json.log",
  "RestartCount": 0,
  "Config": {
    "Tty": false,
    "Labels": [
    ],
    "AttachStdout": false,
    "User": "",
    "Volumes": {
      "\/data": [
      ]
    },
    "AttachStdin": false,
    "AttachStderr": false,
    "OpenStdin": false,
    "Cmd": [
      "\/bin\/sh",
      "-c",
      "\/usr\/sbin\/init.sh"
    ],
    "Image": "sliamb\/paopaodns",
    "WorkingDir": "\/data",
    "Hostname": "acc4366d097f",
    "Env": [
      "CNAUTO=yes",
      "PATH=\/usr\/local\/sbin:\/usr\/local\/bin:\/usr\/sbin:\/usr\/bin:\/sbin:\/bin",
      "TZ=Asia\/Shanghai",
      "UPDATE=weekly",
      "DNS_SERVERNAME=PaoPaoDNS,blog.03k.org",
      "DNSPORT=53",
      "CNFALL=yes",
      "CN_TRACKER=yes",
      "USE_HOSTS=no",
      "IPV6=no",
      "SOCKS5=IP:PORT",
      "SERVER_IP=none",
      "CUSTOM_FORWARD=IP:PORT",
      "AUTO_FORWARD=no",
      "AUTO_FORWARD_CHECK=yes",
      "USE_MARK_DATA=no",
      "RULES_TTL=0",
      "HTTP_FILE=no",
      "QUERY_TIME=2000ms",
      "ADDINFO=no"
    ],
    "Domainname": "",
    "ExposedPorts": {
      "5304\/udp": [
      ],
      "7889\/tcp": [
      ],
      "53\/udp": [
      ],
      "5304\/tcp": [
      ],
      "53\/tcp": [
      ]
    },
    "StdinOnce": false
  },
  "Mounts": [
    {
      "Type": "bind",
      "Source": "\/home\/mydata",
      "Mode": "",
      "Destination": "\/data",
      "RW": true,
      "Propagation": "rprivate"
    }
  ],
  "Id": "acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015",
  "Platform": "linux",
  "HostConfig": {
    "PidMode": "",
    "MemorySwap": 0,
    "ConsoleSize": [
      0,
      0
    ],
    "IOMaximumIOps": 0,
    "DnsOptions": [
    ],
    "CpuPeriod": 0,
    "OomScoreAdj": 0,
    "BlkioWeight": 0,
    "CgroupnsMode": "host",
    "Privileged": false,
    "PortBindings": {
      "53\/udp": [
        {
          "HostIp": "",
          "HostPort": "5335"
        }
      ],
      "53\/tcp": [
        {
          "HostIp": "",
          "HostPort": "5335"
        }
      ]
    },
    "CpuShares": 0,
    "NetworkMode": "default",
    "CpuQuota": 0,
    "DnsSearch": [
    ],
    "NanoCpus": 0,
    "CpuCount": 0,
    "ReadonlyPaths": [
      "\/proc\/bus",
      "\/proc\/fs",
      "\/proc\/irq",
      "\/proc\/sys",
      "\/proc\/sysrq-trigger"
    ],
    "Isolation": "",
    "MaskedPaths": [
      "\/proc\/asound",
      "\/proc\/acpi",
      "\/proc\/kcore",
      "\/proc\/keys",
      "\/proc\/latency_stats",
      "\/proc\/timer_list",
      "\/proc\/timer_stats",
      "\/proc\/sched_debug",
      "\/proc\/scsi",
      "\/sys\/firmware"
    ],
    "Cgroup": "",
    "ContainerIDFile": "",
    "IOMaximumBandwidth": 0,
    "AutoRemove": false,
    "IpcMode": "private",
    "CpuPercent": 0,
    "VolumeDriver": "",
    "OomKillDisable": false,
    "CpuRealtimePeriod": 0,
    "BlkioWeightDevice": [
    ],
    "KernelMemoryTCP": 0,
    "Binds": [
      "\/home\/mydata:\/data"
    ],
    "KernelMemory": 0,
    "MemoryReservation": 0,
    "Devices": [
    ],
    "RestartPolicy": {
      "Name": "always",
      "MaximumRetryCount": 0
    },
    "PublishAllPorts": false,
    "CpusetMems": "",
    "CpusetCpus": "",
    "CpuRealtimeRuntime": 0,
    "CgroupParent": "",
    "ReadonlyRootfs": false,
    "UsernsMode": "",
    "Memory": 0,
    "UTSMode": "",
    "ShmSize": 67108864,
    "LogConfig": {
      "Config": {
        "max-size": "50m",
        "max-file": "2"
      },
      "Type": "json-file"
    },
    "Dns": [
    ],
    "Runtime": "runc"
  },
  "GraphDriver": {
    "Data": {
      "MergedDir": "\/mnt\/sata1-4\/docker\/overlay2\/5f65145a694e230924e36f897e87406e1b47632a292051c7b3af89bead6d9df3\/merged",
      "WorkDir": "\/mnt\/sata1-4\/docker\/overlay2\/5f65145a694e230924e36f897e87406e1b47632a292051c7b3af89bead6d9df3\/work",
      "UpperDir": "\/mnt\/sata1-4\/docker\/overlay2\/5f65145a694e230924e36f897e87406e1b47632a292051c7b3af89bead6d9df3\/diff",
      "LowerDir": "\/mnt\/sata1-4\/docker\/overlay2\/5f65145a694e230924e36f897e87406e1b47632a292051c7b3af89bead6d9df3-init\/diff:\/mnt\/sata1-4\/docker\/overlay2\/f33ee251b68164ca347607ce688e8d498d5650f19952cb332a9674734e1bdc19\/diff:\/mnt\/sata1-4\/docker\/overlay2\/a5834b125223b882944ea4dce8d129ea58b5fe78ebcdd43f575ccb34d5964335\/diff:\/mnt\/sata1-4\/docker\/overlay2\/99c0f0509814fd3563126d59cdd289927ed947e18a9f3c58c967222b126fe309\/diff:\/mnt\/sata1-4\/docker\/overlay2\/4a4d0d1f6ac78f00d666bdf6cfcf716ed7e029d3fbd7399c60ff609b2dbf4d4d\/diff"
    },
    "Name": "overlay2"
  },
  "State": {
    "Pid": 21880,
    "FinishedAt": "0001-01-01T00:00:00Z",
    "StartedAt": "2023-09-20T13:09:15.870819342Z",
    "Error": "",
    "Running": true,
    "Paused": false,
    "OOMKilled": false,
    "Status": "running",
    "ExitCode": 0,
    "Restarting": false,
    "Dead": false
  },
  "Driver": "overlay2",
  "Name": "\/paopaodns",
  "Args": [
    "-c",
    "\/usr\/sbin\/init.sh"
  ],
  "HostnamePath": "\/mnt\/sata1-4\/docker\/containers\/acc4366d097f734e7f4aaeddf89a061289560c401693f427995a18c098981015\/hostname",
  "Created": "2023-09-20T13:09:14.290949024Z",
  "AppArmorProfile": "",
  "Image": "sha256:fc38d02b6f44bf5ecb5ed077e4fa48e1eca0f68417a97f19ce48368b87d86700"
}

**Reply by kkkgo** [reply]
日志中无任何报错，其他选项比如USE_MARK_DATA可以根据文档说明选择性开启。或者进一步配合PaoPaoGateway使用。https://github.com/kkkgo/PaoPaoGateWay

**Reply by eerilqin** [reply]
太感谢你了
